openhtf/util/multicast.py

```diff
@@ -0,0 +1,134 @@
+PING_RESPONSE_KEY = 'OPENHTF_PING_RESPONSE'
+
+
+class MulticastListener(threading.Thread):
+  """Agent that listens (and responds) to short messages on a multicast socket.
+
+  Args:
+    callback: A callable to invoke upon receipt of a multicast message. Will be
+              called with one argument -- the text of the message received.
+              callback can optionally return a string response, which will be
+              transmitted back to the sender.
+    address: Multicast IP address component of the socket to listen on.
+    port: Multicast UDP port component of the socket to listen on.
+    ttl: TTL for multicast messages. 1 to keep traffic in-network.
+  """
+  LISTEN_TIMEOUT_S = 3  # Seconds to listen before retrying.
+
+  def __init__(self,
+               callback,
+               address=DEFAULT_ADDRESS,
+               port=DEFAULT_PORT,
+               ttl=DEFAULT_TTL):
+    super(MulticastListener, self).__init__()
+    self.address = address
+    self.port = port
+    self.ttl = ttl
+    self._callback = callback
+    self._live = False
+    self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    self._sock.setsockopt(socket.IPPROTO_IP,
+                          socket.IP_MULTICAST_TTL,
+                          self.ttl)
+
+  def Stop(self):
+    """Stop listening for messages."""
+    self._live = False
+    self.join()
+
+  def run(self):
+    """Listen for pings until stopped."""
+    self._live = True
+    self._sock.bind(('', self.port))
+    self._sock.settimeout(self.LISTEN_TIMEOUT_S)
+    self._sock.setsockopt(
+        socket.IPPROTO_IP,
+        socket.IP_ADD_MEMBERSHIP,
+        struct.pack(
+            '!4sL',
+            socket.inet_aton(self.address),
+            socket.INADDR_ANY))
+
+    while self._live:
+      try:
+        data, address = self._sock.recvfrom(MAX_MESSAGE_BYTES)
+        _LOG.debug('Received multicast message from %s: %s'% (address, data))
+        response = self._callback(data)
+        if response:
+          self._sock.sendto(response, address)
+      except socket.timeout:
+        continue
+
+
+def send(message,
+         address=DEFAULT_ADDRESS,
+         port=DEFAULT_PORT,
+         ttl=DEFAULT_TTL,
+         timeout_s=2):
+  """Sends a message on the given multicast socket and returns responses.
+
+  Args:
+    message: The string message to send.
+    address: Multicast IP address component of the socket to listen on.
+    port: Multicast UDP port component of the socket to listen on.
+    ttl: TTL for multicast messages. 1 to keep traffic in-network.
+    timeout_s: Seconds to wait for responses.
+
+  Returns: A set of all responses that arrived before the timeout expired.
+           Responses are tuples of (sender_address, message).
+  """
+  result = set()
+  sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+  sock.setsockopt(socket.IPPROTO_IP,
+                  socket.IP_MULTICAST_TTL,
+                  ttl)
+  sock.settimeout(timeout_s)
```

Comment:
I'd put these bits that's shared with the Listener above into a shared function, with a simple (strawman) name like _MakeMulticastSocket(). The settimeout() call above in run() can also be moved into the __init__ making it a candidate for inclusion in the shared function